src/main/scala/is/hail/expr/FunctionRegistry.scala

```diff
@@ -787,15 +792,30 @@ object FunctionRegistry {
   registerMethod("nAlleles", { (x: Variant) => x.nAlleles }, "Number of alleles.")(variantHr(GR), int32Hr)
   registerMethod("isBiallelic", { (x: Variant) => x.isBiallelic }, "True if `v` has one alternate allele.")(variantHr(GR), boolHr)
   registerMethod("nGenotypes", { (x: Variant) => x.nGenotypes }, "Number of genotypes.")(variantHr(GR), int32Hr)
-  registerMethod("inXPar", { (x: Variant) => x.inXPar }, "True if chromosome is X and start is in pseudoautosomal region of X.")(variantHr(GR), boolHr)
-  registerMethod("inYPar", { (x: Variant) => x.inYPar }, "True if chromosome is Y and start is in pseudoautosomal region of Y. *NB: most callers assign variants in PAR to X.*")(variantHr(GR), boolHr)
-  registerMethod("inXNonPar", { (x: Variant) => x.inXNonPar }, "True if chromosome is X and start is not in pseudoautosomal region of X.")(variantHr(GR), boolHr)
-  registerMethod("inYNonPar", { (x: Variant) => x.inYNonPar }, "True if chromosome is Y and start is not in pseudoautosomal region of Y.")(variantHr(GR), boolHr)
+  registerMethodDependentCode("inXPar", { () =>
+    val gr = GR.gr
+    (CM.invokePrimitive1[Any, Any]((x: Any) => x.asInstanceOf[Variant].inXPar(gr)) _).asInstanceOf[Code[Variant] => CM[Code[Boolean]]]
+  }, "True if chromosome is X and start is in pseudoautosomal region of X.")(variantHr(GR), boolHr)
```

Comment:
I think this should be `registerDependentMethod` as before.  `registerMethod` is already going to do the `invokePrimitive` and it will look better from the type perspective.  What's more, you can't do this truly through code because you need to refer to the "Hail compile time" value `gr` that isn't necessary available on the node where this will run so you have to serialize it.

Reply:
@danking Might have comments, he still understands this code better than I do.

Reply:
IIRC, `invokePrimitive1` converts a lambda `T => U` into a `Code[T => U]` by shipping the lambda with the code and passing it as an argument on the worker node.

Reply:
Right.  And `addMethod` wraps that, so I thought it is better to keep using it.

I'm OK with this if you change it to `CM.invokePrimitive1[Variant, Boolean]` and get rid of the cast.  I think that should work (or something close).

Reply:
Jackie tried that at first, the trouble is that `invokePrimitive1` doesn't work correctly with Java Generics. We all want to write:
```scala
CM.invokePrimitive1((x: Variant) => x.inXPar(gr))
```
But when `invokePrimitive1` gets down to `Invokeable.lookupMethod` the class tag is for `Function1` (because we've written a lambda).`Function1` doesn't have a method with the name `apply` and the signature `Variant => Boolean`.

I haven't seriously thought about the right interface here (every time I used it I casted 🤷‍♂️ I guess when I was writing it I had in mind the leaked abstraction not the true abstraction).

1. Maybe the simple thing is `invokePrimitive1(generic=True, f)` and `generic=True` assumes that all arguments and the return type are generic. That works for `Function1` but not for other methods.

2. Another option is to modify `Invokeable.lookupMethod` to specially check for `Function1` and friends and know that their parameters are generic.

3. One fully "correct" answer to this is to exhaustively specify the type signature to `invokePrimitive*` marking every instance of a generic type as such.

4. A costly and hacking solution, is to iteratively try all the possible "generifications" of a signature looking for a method that matches.

I think I rather like 2.

Reply:
The heart of the issue is asm4s cannot invoke methods that have generics correctly. We need to a) find the right method, and b) cast the result, if it was generic.

Reply:
> The heart of the issue

Good!  Go to the heart of the issue.  ALWAYS go to the heart of the issue.

The worst technical debt is debt that comes from not going to the heart of the issue.  Just making local fixes to get past code review or fix a bug, without thinking about the larger conceptual framework that this code is in conflict with, or the error in the larger conceptual framework that allowed this problem to pop up in the first place.  I can't overstate how valuable I think this habit is.  When you do this, it is easier to make fast decisions because they're naturally driven by the conceptual framework and there is a clear right answer.  When it isn't easy, that makes you need to take a step back and think hard about how to revise the framework.  If I've been more productive than the average programmer, this is 100% why.

You say:

 > We all want to write:
 >     CM.invokePrimitive1((x: Variant) => x.inXPar(gr))

That is clearly the right answer so let's figure out how to do that.  That rules out 3.

Special casing FunctionN clearly isn't right either.  That rules out 1, 2.

4: "hacking".  Nope, that's out, too.  Let's try again.

Reply:
> `Function1` doesn't have a method with the name apply and the signature `Variant => Boolean`.

Matching happens on arguments.  It does have a signature that takes `Object`.  `Object` is a superclass of `Variant`.  So it should match.  I don't understand this.

Reply:
@jigold @danking In the mean time, there's a question of what @jigold should do.  I think she should write `reigsterDependentMethod` which has the correct type interaction and already hides the hacky cast internally.  In other words, the previous non-code version was fine, it just needs to be made dependent.  Sound good?

Reply:
I don't understand your comment about matching.

As I understand it, the [reflection API looks for a method with a given signature](https://docs.oracle.com/javase/7/docs/api/java/lang/Class.html#getMethod(java.lang.String,%20java.lang.Class...)). asm4s constructs a signature from the class tags it is given, in this case: `Variant` and `Boolean`. The resulting signature is `apply(Lis/hail/variant/Variant;)Ljava/lang/Boolean;`, `Function1` doesn't have a method matching that signature. AFAIK, all of this is at the JVM byte code level, these notions of generic parameters and the like are completely gone.

The way to properly resolve this issue is to sit down with Scala's reflection API and figure out how to preserve type information all the way to `lookupMethod`. I suspect this is easy: we need `TypeTag` instead of `ClassTag`.

From that point forward, I have no concrete plan. Somehow we need to recreate the Java method matching process. `TypeTag` [is rather confusingly documented](http://www.scala-lang.org/api/2.11.1/scala-reflect/index.html#scala.reflect.api.TypeTags$TypeTag) and it's not clear how to get a list of non-erased methods. If we can get that list, then we just have to re-implement Java-style method matching. If we're lucky Scala already has such a matcher implemented on `TypeTag`.

Alternatively, perhaps there's a way to get the Java Reflection hierarchy's `Type` and try that route.

I basically thrashed around trying to figure out how to get all this to work a while back and couldn't find a reasonable solution, so I left it the way it is, which more or less works if you think of `Code[T]` as representing byte code, not java code. You have to manually write the casts around generic parameters. I realize this isn't ideal, we should fix it, but we need to set the priority of fixing this relative to other fixing/working on other things.